### PR TITLE
Feature: Create a directory and file with Pathname#touch

### PR DESCRIPTION
**Pull request by schneems**
Right now if a developer wants to create a file and is not sure if the path exists yet or not they must:

```ruby
Pathname.new("/a/b/c/d.txt").tap {|p| p.dirname.mkpath; FileUtils.touch(p)}
```

After this patch a developer can instead call:

```ruby
Pathname.new("/a/b/c/d.txt").touch
```

An alternative name for this behavior could be `mkfile` but I think it is confusing to have a `mkfile` and a `mkpath` where one creates a directory and one creates a file.

### CODE REVIEW

**Comment by marcandre** [comment]
I understand how the `mkpath` is necessary, but not the `touch`. `File.write` / `File.open` will create the file if it's not there, no?

**Comment by schneems** [comment]
Here’s one place I use it https://github.com/heroku/heroku-buildpack-ruby-experimental/blob/7afbb4401600cef9c8d85371e7c84baf817568a6/lib/heroku_buildpack_ruby/metadata/cnb.rb#L21

In this case the file might be empty but I want to be able to read it. There are also other cases where I just want the file to exist or I want to create it and pass it to some other interface that expects it to already be created.

This could fail if the file doesn’t exist:

    Pathname.new(“foo.txt”).read

But this would guarantee that it would be present:

    Pathname.new(“foo.txt”).touch.read

Basically I would use it anywhere I would otherwise use “touch” if I were shell scripting.

**Comment by simi** [comment]
@schneems I see this is second request to change in stdlib because of your problem. I have faced similar problems using stdlib since some parts are really not in line with Ruby style of 2020. It would be great to revisit some of those "hot" places before 3.0 release. Would you mind better to open redmine ticket to discuss all potential places to changes in there? I got more examples to add as well (regarding Tempfile for example).

File: ext/pathname/lib/pathname.rb [hunk 1]
```diff
@@ -585,6 +585,27 @@ def mkpath
...
+  #   Dir.exist?("/a/b/c") # => false
+  #
+  #   p = Pathname.new("/a/b/c/d.txt")
+  #   p.file? => false
+  #   p.touch
+  #   p.file? => true
+  #
+  #   Dir.exist?("/a/b/c") # => true
+  def touch
+    require 'fileutils'
```
Comment (simi) [inline ext/pathname/lib/pathname.rb:602]: :thinking: If I understand it well, "fileutils" was "promoted" to gem and Ruby vendored code can't rely on its presence now.

**Comment by schneems** [comment]
> Would you mind better to open redmine ticket to discuss all potential places to changes in there? I got more examples to add as well

I think it's generally better to keep requests small so they can be individually accepted or rejected. I'm not sure how to get changes accepted into Pathname. I already opened one ticket here https://bugs.ruby-lang.org/issues/17287. I see someone thumbs down your comment but I don't know why.

>  I got more examples to add as well (regarding Tempfile for example).

I think writing each up as it's own ticket for it's own discussion would be valuable. If you also have a patch people can use as reference it likely helps.

File: ext/pathname/lib/pathname.rb (hunk 1 above)
Comment (marcandre) [reply ext/pathname/lib/pathname.rb:602]: my understanding is that all builtin libraries will still be present no matter what, either within Ruby or as preinstalled gem.

**Comment by schneems** [comment]
https://bugs.ruby-lang.org/issues/17295

**Comment by hsbt** [comment]
This PR was rejected at https://bugs.ruby-lang.org/issues/17295#note-9
